File: activities/threesum/threesum.c
// Measure the time its take to compute some function.

#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

/**
 * Return the number seconds since the Epoch.
 * @returns a number of seconds
 */
double time_get() {
    struct timeval t;
    gettimeofday(&t, NULL);
    return t.tv_sec + t.tv_usec/1000000.;
}

/**
 * Compute the number of distinct triples such that xs[i] + xs[j] + xs[k] = 0
 * @param n  the number of elements in `xs`
 * @param xs  an array of integers
 * @returns the number of distinct triples that sum to zero
 */
int count_triples(int n, int xs[]) {
    int count = 0;
    for (int i = 0; i < n; ++i) {
        for (int j = i + 1; j < n; ++j) {
            for (int k = j + 1; k < n; ++k) {
	              if (xs[i] + xs[j] + xs[k] == 0) {
	                  count += 1;
	              }
            }
        }
    }
    return count;
}

/**
 * Run a doubling experiment.
 * @param n  problem size
 * @returns the number of seconds to run the experiment
 */
double time_count_triples(int n) {
    int *xs = malloc(n * sizeof(int));
    if (xs == NULL) {
        exit(1);
    }
    for (int i = 0; i < n; ++i) {
        xs[i] = lrand48();
    }
    double start = time_get();
    count_triples(n, xs);
    double end = time_get();
    free(xs);
    return end - start;
}

int main() {
    int n = 16;
    double t_prev = time_count_triples(n);
    printf("    size     time    ratio\n");
    for (;;) {
        n *= 2;
        double t = time_count_triples(n);
        printf("%8d %8.1f %8.2f\n", n, t, t/t_prev);
        t_prev = t;
    }
}

File: activities/threesum/threesum.py
"""Measure the time its take to compute some function."""

import random
import time

def count_triples(xs : list[int]) -> int:
    """Return the number of distinct triples (i, j, k) such that
    xs[i] + xs[j] + xs[k] = 0"""
    count = 0
    n = len(xs)
    for i in range(n):
        for j in range(i + 1, n):
            for k in range(j + 1, n):
                if xs[i] + xs[j] + xs[k] == 0:
                    count += 1
    return count

def time_count_triples(n : int) -> float:
    """Run a doubling experiment."""
    xs = [ random.randint(-1000000, +1000000) for _ in range(n)]
    start = time.time()
    count_triples(xs)
    end = time.time()
    return end - start

def main():
    n = 16
    t_prev = time_count_triples(n)
    print('{:>8s} {:>8s} {:>8s}'.format('size', 'time', 'ratio'))
    while True:
        n *= 2
        t = time_count_triples(n)
        print('{:8d} {:8.1f} {:8.2f}'.format(n, t, t/t_prev))
        t_prev = t

if __name__ == '__main__':
    main()
